digiroad2-geo/src/main/scala/fi/liikennevirasto/digiroad2/linearasset/PavedRoadFiller.scala

```diff
@@ -5,31 +5,29 @@ import fi.liikennevirasto.digiroad2.linearasset.LinearAssetFiller._
 
 object PavedRoadFiller extends AssetFiller {
 
-  //RoadLinks should contain only 1 pavement asset. If multiple are present, redundant assets should be expired
-  override def adjustAssets(roadLink: RoadLinkForFillTopology, linearAssets: Seq[PieceWiseLinearAsset], changeSet: ChangeSet): (Seq[PieceWiseLinearAsset], ChangeSet) = {
-    val (filteredAssets, updatedChangeSet) = (linearAssets.size > 1) match {
-      case true =>
-        val (redundantAssets, validAssets) = linearAssets.partition(asset => isRedundantUnknownPavementAsset(asset, roadLink))
-        if (validAssets.nonEmpty) {
-          logger.info(s"LinearAsset ${redundantAssets.map(_.id)} removed from process due to having Unknown PavementClass and not filling whole RoadLink")
-          (validAssets, changeSet.copy(expiredAssetIds = changeSet.expiredAssetIds ++ redundantAssets.map(_.id).toSet))
-        } else (linearAssets, changeSet) // return all unknowns if no valid assets.
-      case false => (linearAssets, changeSet)
+  private def expireReplacedUnknownPavementAssets(assetsIdsToExpire: Set[Long], changeSet: ChangeSet) = {
+    if (assetsIdsToExpire.nonEmpty) {
+      logger.info(s"LinearAssets ${assetsIdsToExpire.mkString(", ")} expired due to Unknown PavementClass and being replaced by asset with known PavementClass")
     }
-    super.adjustAssets(roadLink, filteredAssets, updatedChangeSet)
+    changeSet.copy(expiredAssetIds = changeSet.expiredAssetIds ++ assetsIdsToExpire)
   }
 
-  /** *
-   * Filters out Pavement assets that do not fill the whole RoadLink and have an Unknown value
-   *
-   * @param asset
-   * @param roadLink
-   * @return
+  /**
+   * Filter out assets with Unknown PavementClass if any other PavementClass is present.
+   * If all assets have Unknown PavementClass, do nothing.
+   * @param roadLink  which we are processing
+   * @param linearAssets  assets on link
+   * @param changeSet record of changes for final saving stage
+   *  @return assets and changeSet
    */
-  private def isRedundantUnknownPavementAsset(asset: PieceWiseLinearAsset, roadLink: RoadLinkForFillTopology): Boolean = {
-    val assetLength = asset.endMeasure - asset.startMeasure
-    val linkLength = roadLink.length
-    val pavementClass = PavementClass.extractPavementClass(asset.value).getOrElse(DynamicPropertyValue(""))
-    (assetLength - linkLength).abs > MaxAllowedMValueError && PavementClass.applyFromDynamicPropertyValue(pavementClass.value) == PavementClass.Unknown 
+  override def adjustAssets(roadLink: RoadLinkForFillTopology, linearAssets: Seq[PieceWiseLinearAsset], changeSet: ChangeSet): (Seq[PieceWiseLinearAsset], ChangeSet) = {
```

Comment:
Tässä, kun ajetaan fillerillä, niin pitää ottaa huomioon, et tää ajetaan myös UI:ta käytettäessä. Oletko katsonut, että ei tule sivuvaikutuksia? Olisiko tämä logiikka mahdollista tehdä kokonaan updaterin puolella?

Reply:
Erittäin hyvä nosto. Täytyy tarkistaa.